### ISSUE TITLE
BootstrapAgent::sled_reset doesn't fully reset the sled

### ISSUE THREAD

**Issue by jgallagher** [oxidecomputer/omicron#3820]
The bootstrap agent exposes an API to reset the sled, and it notes its intent: https://github.com/oxidecomputer/omicron/blob/80b5f7f2f9253695f13832c62dface4116dbb2e8/sled-agent/src/bootstrap/agent.rs#L983-L984

However, it doesn't currently fully reset the state:

1. There's no way to "uninitialize" the LRTQ salt which gets set in `request_sled_agent`. (See #3818 - once `request_sled_agent` has been called once with a rack_id, all subsequent calls must provide the same rack_id).
2. (I think?) It does not unadvertise the sled prefix that is advertised in `request_sled_agent`.
3. (Unsure) Does it clean up any configuration it has sent to the switches (e.g., NAT entries for NTP/external DNS/instance zones)?

I chatted with @andrewjstone briefly and he suggested we may want to do something more extreme: remove anything written to the M.2s (bootstore / ledger), still remove persistent storage as `reset_sled` does today, then _reboot the sled_. That would take care of in-memory caches like LRTQ (unsure if there are more of these I've missed); I'm not sure about undoing switch configurations.

**Comment by davepacheco** [comment]
Along those lines, I've seen success (in different systems, long ago at this point) with an approach where the system:

- on first boot, takes a ZFS snapshot of the root ZFS dataset on the root pool (probably the M.2s, in our case)
- to factory reset, rollback to the snapshot and reboot

That's probably not quite enough here but I think it can take care of quite a lot of state in a high-confidence way.

**Comment by iliana** [comment]
In the 5/28/24 update sync / reconfigurator meeting we discussed rack reset, among other related topics. There's not much new information there that's not in the issue description here, other than the suggestion that instead of a reboot, sled-agent could exec a mini-sled-agent (or re-exec itself in some sort of reduced fashion) to finish up any rack reset tasks and let Wicket know that it finished.

(The related topics included allowing a cooperative sled-agent to install OS updates on wicket's behalf to significantly reduce the amount of time spent waiting on installinator during dev/test cycles on racklettes.)

In any of these cases, doing this "properly" means sled-agent has to learn how to "park" itself, and there's a _lot_ of work to be done here. https://github.com/oxidecomputer/omicron/issues/3718 has some discussion, as do the 8/4/23 and 8/8/23 control plane huddle meetings, but we seem to have consensus that this is not a necessary step toward Nexus-driven upgrade.

There's ways to do it that are sufficient for dev/test workflows though, and we discussed a few different ideas for half-implementing a parked state. Each of them had the drawback of implementing some change that we would eventually need to fix or rewrite for the automated upgrade case. So overall I don't think there's a lot of appetite to fix rack reset until either more of the automated upgrade story is in place, or until we absolutely must have it because we can no longer run the clean-slate script on the racklettes.

I'm also not certain if "rack reset" is intended to become a customer-accessible factory reset feature (I assume so?) and/or a "secure erase" feature for decommissioning. (It's neither of these cases today, and with the warning added in #6493 I'm less concerned about customers accidentally thinking it is.)
